server/src/core/distribution-report/distribution-report-service.ts

```diff
@@ -0,0 +1,98 @@
+
+  constructor(db: Db, args: DistributionReportArgs) {
+    this.db = db;
+    this.collection = db.collection(COLLECTION);
+    this.args = args;
+  }
+
+  async sendDistributionReportsToDonors(): Promise<void> {
+    try {
+      const reports: DistributionReport[] = await this.args.transactions.generateDistributionReportDocs();
+      await this.sendDistributionReportMessages(reports);
+      if (reports.length) {
+        await this.collection.insertMany(reports);
+      }
+    }
+    catch (e) {
+      console.error("Error occured: ", e.message);
+      rethrowIfAppError(e);
+      throw createDbOpFailedError(e.message);
+    }
+  }
+
+  private async sendDistributionReportMessages(reportDocs: DistributionReport[]): Promise<void> {
+    try {
+      reportDocs.forEach(async (report: DistributionReport) => {
+        const donor = await this.args.users.getById(report.donor);
+        const beneficiaries = await this.getBeneficiaries(report.beneficiaries);
+
+        const messages = this.getSmsAndEmailMessages(report, donor, beneficiaries);
+
+        await Promise.all([
+          this.args.smsProvider.sendSms(donor.phone, messages.sms),
+          this.args.emailProvider.sendEmail(donor.email, messages.email),
+        ]);                 
+      });
+    }
+    catch(error) {
+      console.error('Error occurred when handling event', error);
+    }
+  }
+
+  private getSmsAndEmailMessages(report: DistributionReport, donor: User, beneficiaries: User[]): SmsAndEmailMessages {
+    const sms = `Hello ${donor.name},
+                \nIn the last 24 hours, Ksh ${report.totalDistributedAmount} has been transferred 
+                from your SocialRelief donation to the following beneficiaries:
+                ${this.beneficiariesAndAmountReceived(beneficiaries, report.receivedAmount)}
+                \nOn behalf of all beneficiaries, we at SocialRelief say thank you!
+                \nIf you wish to donate more, click ${this.generateDonateLink(donor, report.totalDistributedAmount)}`;
+
+    const email = `<p>
+                  Hello ${donor.name},<br><br>
+                  In the last 24 hours, <strong>Ksh ${report.totalDistributedAmount}</strong> has been transferred
+                  from your SocialRelief donation to the following beneficiaries:
+                  ${this.beneficiariesAndAmountReceived(beneficiaries, report.receivedAmount, false)}<br>
+                  On behalf of all beneficiaries, we at SocialRelief say thank you!<br>
+                  If you wish to donate more, click ${this.generateDonateLink(donor, report.totalDistributedAmount)}
+                  </p>`;
+
+    return { sms, email };
+  }
+
+  private generateDonateLink(user: User, totalDistributedAmount: number): string {
+    const amount: number = totalDistributedAmount > 2000 ? totalDistributedAmount : 2000;
+    return `<a href='socialrelief.co?donate=true&n=${user.name}&e=${user.email}&p=${user.phone}&a=${amount}'>here</a>`;
+  }
```

Comment:
I think this should be a utility function or a function in a link generation service. I think it should just generate the plain link, not a html anchor element. The email messages could create the anchor element and wrap the link. Should we hardcode socialrelief.co? Or should we use the `webappBaseUrl` config variable?

If we had a link generation service, I think that could help pass it the necessary config, e.g.

```ts
const linkGenerator = new LinkGenerator({ baseUrl: config.webappBaseUrl });

linkGenerator.getUserDonateLink(user, amount);
linkGenerator.getDonateLink({ name, email, phone, amount });
```
But I think it would be great if the donate link.

A user of the system has advised us not to use long links in SMS messages. Maybe we could consider integrating the bitly service via its API to shorten links, especially or SMS. But if this is a lot of work. We could do it in the next version.